allofplos/corpus_class.py

```diff
@@ -0,0 +1,267 @@
+
+    def hashtable_to_json(self, hashtable=None, overwrite=True):
+        """Dump `self.hashtable` OrderedDict into a .json file.
+
+        :param hashtable: the hashtable, defaults to None
+        :param overwrite: whether to replace existing .json file, defaults to True"""
+        if hashtable is None:
+            hashtable = self.hashtable()
+        if overwrite is True or os.path.isfile(hash_json) is False:
+            with open(hash_json, 'w') as fp:
+                json.dump(hashtable, fp)
+        else:
+            print("aborting: .json hashtable exists. To overwrite, set parameter 'overwrite' to False.")
+        return hashtable
+
+    def read_hashtable(self):
+        """Read the `hash_json` file into memory.
+
+        Also see `self.hashtable{}`
+        """
+        if os.path.isfile(hash_json):
+            return OrderedDict(json.load(open(hash_json)))
+        else:
+            print("Hashtable not found; create with Corpus().hashtable()")
+            return None
+
+    def update_hashtable(self, update_json=True):
+        """Update the hashtable by comparing the hashtable .json file and XML dates modified.
+
+        This is to incrementally update the hashtable but timestamps are not foolproof. When in doubt,
+        create from scratch using `self.hashtable_to_json`.
+        :param update_json: whether to also update the accompanying json file
+        :return: updated hashtable for new files
+        """
+        hashtable = self.read_hashtable()
+        if hashtable:
+            json_mod_date = datetime.datetime.fromtimestamp(os.path.getmtime(hash_json))
+            modded_files = [fname for fname in self.filepaths if json_mod_date < datetime.datetime.fromtimestamp(os.path.getmtime(fname))]
+            print(len(modded_files))
+
+            # get new hashes, and replace or insert their values into hashtable
+            for fname in tqdm(modded_files):
+                file_hash = hash_file(fname)
+                if hashtable.get(fname, '') != file_hash:
+                    hashtable[filename_to_doi(fname)] = file_hash
+                else:
+                    pass
+
+            # re-order hashtable dict by DOI/key:
+            resorted_hashtable = OrderedDict(sorted(hashtable.items(), key=lambda t: t[0]))
+
+        if update_json:
+            json_dump = self.hashtable_to_json(hashtable=resorted_hashtable)
+
+        else:
+            print('Hashtable .json file not found. Create with `Corpus().hashtable_to_json()`')
+            resorted_hashtable = {}
+
+        return resorted_hashtable
+
+    # @symlinks.setter
+    # def symlinks(self, value):
+    #     """Sets a corpus object using a set of dois.
+
+    #     Converts a filename to DOI using an existing function.
+    #     :param value: filename
+    #     :type value: string
+    #     """
+    #     self.doi = filename_to_doi(value)
+
+    @classmethod
+    def create(cls, directory=corpusdir, overwrite=False):
```

Comment:
I'm not 100% sure that this should be part of the initial corpus PR… it seems like this is a really useful but potentially complciated feature… it could be part of the first PR that is merged about the corpus after the corpus itself has been a bit more battle tested.